nemo/collections/nlp/models/language_modeling/megatron_lm_adapter_encoder_decoder_model.py

```diff
@@ -0,0 +1,166 @@
+                logging.info(f'Adapters weights loaded successfully from {self.cfg.adapters_file}')
+            self.loading_base_model = True
+
+    def _validate_adapters_cfg(self, cfg):
+        assert cfg.type in ['parallel_adapter', 'linear_adapter']
+        assert hasattr(cfg, 'adapter_dim')
+        assert cfg.norm_position in ['pre', 'post']
+
+        if hasattr(cfg, 'norm_type'):
+            assert cfg.norm_type in ['mixedfusedlayernorm', 'layernorm']
+
+        for val in ['row_init_method', 'column_init_method']:
+            if hasattr(cfg, val):
+                assert cfg.get(val) in ['xavier', 'zero', 'normal']
+
+    def _state_dict(self, destination=None, prefix=None, keep_vars=False):
+        state_dict_ = {}
+
+        for name, module in self.enc_dec_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    adapter_module = module.adapter_layer[adapter_key]
+                    state_adapter_key = ':'.join([name, adapter_key])
+                    state_dict_[state_adapter_key] = adapter_module.state_dict()
+        return state_dict_
+
+    def _load_adapters_weights(self, state_dict, strict=True):
+        if not self.loading_base_model:
+            # load adapters only
+            for name, module in self.enc_dec_model.named_modules():
+                if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                    for adapter_key in self.adapter_name_keys:
+                        adapter_module = module.adapter_layer[adapter_key]
+                        state_adapter_key = ':'.join([name, adapter_key])
+                        # only load the adapters if they are in the state_dict
+                        if state_adapter_key in state_dict:
+                            adapter_module.load_state_dict(state_dict[state_adapter_key], strict)
+        else:
+            # load the base model once and change the loading to adapters only
+            super().load_state_dict(state_dict, strict=False)
+            self.loading_base_model = False
+
+    def setup_optimizer_param_groups(self):
+        """
+        ModelPT override. Optimizer will get self._optimizer_param_groups. 
+        Makes two optimizer param groups, one for the frozen model params
+        and one for the prompt-table/prompt-encoder params. The learning 
+        rate for the frozen model's params will always be zero effectively
+        freezing the model's params but still allowing for the needed gradients
+        to be passed around in pipeline parallel models. The prompt-encoder 
+        and/or prompt table will use the learning rate set by the user. 
+        """
+        # self freeze and unfreeze enabled adapters
+        if hasattr(self.cfg, 'adapter_tuning'):
+            self.freeze()
+            param_groups = {'params': [p for p in self.enc_dec_model.parameters()]}
```

Comment:
Perhaps for another PR it would be nice to only have the adapter params in param_groups like here: https://github.com/NVIDIA/NeMo/pull/5086